python/lsst/daf/persistence/butler.py

```diff
@@ -822,11 +828,48 @@ def dataRef(self, datasetType, level=None, dataId={}, **rest):
         return ButlerDataRef(subset, subset.cache[0])
 
     def _read(self, location):
-        self.log.debug("Starting read from %s", location)
-        results = location.repository.read(location)
-        if len(results) == 1:
-            results = results[0]
-        self.log.debug("Ending read from %s", location)
+        """Unpersist an object using data inside a butlerLocation object.
+
+        A weakref to loaded objects is cached here. If the object specified by the butlerLocaiton has been
+        loaded before and still exists then the object will not be re-read. A ref to the already-existing
+        object will be returned instead.
+
+        Parameters
+        ----------
+        location - ButlerLocation
+            A butlerLocation instance populated with data needed to read the object.
+
+        Returns
+        -------
+        object - an instance of the object specified by the butlerLoction.
+            The object specified by the butlerLocation will either be loaded from persistent storage or will
+            be fetched from the object cache (if it has already been read before).
+        """
+        def hasher(butlerLocation):
+            """Hash a butler location for use as a key in the object cache."""
+            try: # is the dataId a daf_persistence.Policy?
```

Comment:
How does usedDataId (which looks to be documented as a dict and created as a new dict each time) get to be a Policy?